File: sdk-go/crypto.go
package tuish

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"
)

var (
	ErrInvalidFormat    = errors.New("invalid license format")
	ErrInvalidSignature = errors.New("invalid signature")
	ErrExpired          = errors.New("license expired")
	ErrMachineMismatch  = errors.New("machine mismatch")
)

// VerifyResult contains the result of license verification.
type VerifyResult struct {
	Valid   bool
	Payload *LicensePayload
	Reason  LicenseInvalidReason
}

// ParsedLicense contains the parsed components of a license string.
type ParsedLicense struct {
	Header    LicenseHeader
	Payload   LicensePayload
	Signature []byte
	RawHeader string
	RawPayload string
}

// base64URLDecode decodes a base64url-encoded string.
func base64URLDecode(s string) ([]byte, error) {
	// Add padding if needed
	switch len(s) % 4 {
	case 2:
		s += "=="
	case 3:
		s += "="
	}

	// Convert from base64url to standard base64
	s = strings.ReplaceAll(s, "-", "+")
	s = strings.ReplaceAll(s, "_", "/")

	return base64.StdEncoding.DecodeString(s)
}

// ParseLicense parses a license string into its components.
func ParseLicense(licenseString string) (*ParsedLicense, error) {
	parts := strings.Split(licenseString, ".")
	if len(parts) != 3 {
		return nil, ErrInvalidFormat
	}

	headerB64, payloadB64, signatureB64 := parts[0], parts[1], parts[2]
	if headerB64 == "" || payloadB64 == "" || signatureB64 == "" {
		return nil, ErrInvalidFormat
	}

	// Decode header
	headerBytes, err := base64URLDecode(headerB64)
	if err != nil {
		return nil, fmt.Errorf("decode header: %w", err)
	}

	var header LicenseHeader
	if err := json.Unmarshal(headerBytes, &header); err != nil {
		return nil, fmt.Errorf("parse header: %w", err)
	}

	// Decode payload
	payloadBytes, err := base64URLDecode(payloadB64)
	if err != nil {
		return nil, fmt.Errorf("decode payload: %w", err)
	}

	var payload LicensePayload
	if err := json.Unmarshal(payloadBytes, &payload); err != nil {
		return nil, fmt.Errorf("parse payload: %w", err)
	}

	// Decode signature
	signature, err := base64URLDecode(signatureB64)
	if err != nil {
		return nil, fmt.Errorf("decode signature: %w", err)
	}

	return &ParsedLicense{
		Header:     header,
		Payload:    payload,
		Signature:  signature,
		RawHeader:  headerB64,
		RawPayload: payloadB64,
	}, nil
}

// ParsePublicKey parses a public key from SPKI base64 or hex format.
// Returns the raw 32-byte key.
func ParsePublicKey(publicKey string) (ed25519.PublicKey, error) {
	// Check if it's SPKI base64 format (starts with MCow for Ed25519)
	if strings.HasPrefix(publicKey, "MCow") || strings.HasPrefix(publicKey, "MCoq") {
		decoded, err := base64.StdEncoding.DecodeString(publicKey)
		if err != nil {
			return nil, fmt.Errorf("decode SPKI key: %w", err)
		}

		// SPKI format: 12 byte header + 32 byte key
		if len(decoded) != 44 {
			return nil, fmt.Errorf("invalid SPKI key length: expected 44 bytes, got %d", len(decoded))
		}

		// Extract the raw key (last 32 bytes)
		return ed25519.PublicKey(decoded[12:]), nil
	}

	// Check if it's hex format (64 characters = 32 bytes)
	if len(publicKey) == 64 {
		key, err := hex.DecodeString(publicKey)
		if err != nil {
			return nil, fmt.Errorf("decode hex key: %w", err)
		}
		return ed25519.PublicKey(key), nil
	}

	return nil, errors.New("invalid public key format: expected SPKI base64 (MCow...) or 64-character hex")
}

// VerifyLicense verifies a license signature and checks expiration/machine binding.
func VerifyLicense(licenseString string, publicKey ed25519.PublicKey, machineID string) *VerifyResult {
	parsed, err := ParseLicense(licenseString)
	if err != nil {
		return &VerifyResult{Valid: false, Reason: ReasonInvalidFormat}
	}

	// Verify signature
	message := []byte(parsed.RawHeader + "." + parsed.RawPayload)
	if !ed25519.Verify(publicKey, message, parsed.Signature) {
		return &VerifyResult{Valid: false, Reason: ReasonInvalidSignature}
	}

	// Check expiration
	if parsed.Payload.ExpiresAt != nil && *parsed.Payload.ExpiresAt < time.Now().UnixMilli() {
		return &VerifyResult{Valid: false, Payload: &parsed.Payload, Reason: ReasonExpired}
	}

	// Check machine ID if license is locked to a specific machine
	// If license.MachineID is nil, license is valid on any machine
	if machineID != "" && parsed.Payload.MachineID != nil && *parsed.Payload.MachineID != machineID {
		return &VerifyResult{Valid: false, Payload: &parsed.Payload, Reason: ReasonMachineMismatch}
	}

	return &VerifyResult{Valid: true, Payload: &parsed.Payload}
}

// ExtractLicensePayload extracts the payload from a license without verification.
// This is for display purposes only - never trust unverified payloads.
func ExtractLicensePayload(licenseString string) (*LicensePayload, error) {
	parsed, err := ParseLicense(licenseString)
	if err != nil {
		return nil, err
	}
	return &parsed.Payload, nil
}

// IsLicenseExpired checks if a license is expired based on payload.
// Does not verify the signature.
func IsLicenseExpired(licenseString string) bool {
	payload, err := ExtractLicensePayload(licenseString)
	if err != nil {
		return true
	}
	if payload.ExpiresAt == nil {
		return false // Perpetual license
	}
	return *payload.ExpiresAt < time.Now().UnixMilli()
}

File: py/tuish/crypto.py
"""Ed25519 signature verification using PyNaCl."""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Literal

from nacl.exceptions import BadSignatureError
from nacl.signing import VerifyKey

from tuish.exceptions import TuishCryptoError
from tuish.models import LicenseHeader, LicensePayload, SignedLicense
from tuish.utils import bytes_to_json, from_base64url, hex_to_bytes

# SPKI header for Ed25519 public keys (12 bytes)
# Format: 30 2a 30 05 06 03 2b 65 70 03 21 00 [32 bytes key]
ED25519_SPKI_HEADER = "MCowBQYDK2VwAyEA"


VerifyReason = Literal[
    "invalid_format",
    "invalid_signature",
    "expired",
    "machine_mismatch",
]


@dataclass
class VerifyResult:
    """Result of license verification."""

    valid: bool
    payload: LicensePayload | None = None
    reason: VerifyReason | None = None


def parse_public_key(public_key: str) -> bytes:
    """
    Parse public key from SPKI base64 or hex format.
    Returns raw 32-byte key.
    """
    # Check if SPKI base64 format
    if public_key.startswith(ED25519_SPKI_HEADER) or public_key.startswith("MCoq"):
        import base64
        decoded = base64.b64decode(public_key)

        # SPKI format: 12 byte header + 32 byte key
        if len(decoded) != 44:
            raise TuishCryptoError(
                f"Invalid SPKI public key length: expected 44 bytes, got {len(decoded)}"
            )

        # Extract raw key (last 32 bytes)
        return decoded[12:]

    # Check if hex format (64 characters = 32 bytes)
    if len(public_key) == 64:
        try:
            return hex_to_bytes(public_key)
        except ValueError:
            pass

    raise TuishCryptoError(
        "Invalid public key format. Expected SPKI base64 (MCow...) or 64-character hex"
    )


def parse_license(license_string: str) -> SignedLicense | None:
    """
    Parse a license string into its components.
    Returns None if format is invalid.
    """
    try:
        parts = license_string.split(".")
        if len(parts) != 3:
            return None

        header_b64, payload_b64, signature_b64 = parts

        if not all([header_b64, payload_b64, signature_b64]):
            return None

        # Decode header and payload
        header_json = bytes_to_json(from_base64url(header_b64))
        payload_json = bytes_to_json(from_base64url(payload_b64))

        return SignedLicense(
            header=LicenseHeader(**header_json),
            payload=LicensePayload(**payload_json),
            signature=signature_b64,
        )
    except Exception:
        return None


def verify_license(
    license_string: str,
    public_key_hex: str,
    machine_id: str | None = None,
) -> VerifyResult:
    """
    Verify a license signature and check validity.

    Args:
        license_string: The full license string (header.payload.signature)
        public_key_hex: 64-character hex string of public key
        machine_id: Optional machine fingerprint to verify binding

    Returns:
        VerifyResult with valid=True/False and reason if invalid
    """
    # Parse the license
    parsed = parse_license(license_string)
    if not parsed:
        return VerifyResult(valid=False, reason="invalid_format")

    # Prepare verification data
    parts = license_string.split(".")
    message = f"{parts[0]}.{parts[1]}".encode()
    signature_bytes = from_base64url(parsed.signature)
    public_key_bytes = hex_to_bytes(public_key_hex)

    # Verify signature using PyNaCl
    try:
        verify_key = VerifyKey(public_key_bytes)
        # PyNaCl expects signature + message concatenated for verify()
        # But we can use verify_key.verify(smessage) where smessage = sig + msg
        # Or: construct signed message manually
        verify_key.verify(message, signature_bytes)
    except BadSignatureError:
        return VerifyResult(valid=False, reason="invalid_signature")
    except Exception:
        return VerifyResult(valid=False, reason="invalid_signature")

    payload = parsed.payload

    # Check expiration (exp is in milliseconds)
    now_ms = int(time.time() * 1000)
    if payload.exp is not None and payload.exp < now_ms:
        return VerifyResult(valid=False, payload=payload, reason="expired")

    # Check machine ID binding
    # If license.mid is empty string, license is valid on any machine
    if machine_id and payload.mid and payload.mid != machine_id:
        return VerifyResult(valid=False, payload=payload, reason="machine_mismatch")

    return VerifyResult(valid=True, payload=payload)


def extract_license_payload(license_string: str) -> LicensePayload | None:
    """Extract payload from license without verification (for display only)."""
    parsed = parse_license(license_string)
    return parsed.payload if parsed else None


def is_license_expired(license_string: str) -> bool:
    """Check if license is expired based on payload (without signature verification)."""
    payload = extract_license_payload(license_string)
    if not payload:
        return True
    if payload.exp is None:
        return False  # Perpetual license
    now_ms = int(time.time() * 1000)
    return payload.exp < now_ms


def get_license_time_remaining(license_string: str) -> int | None:
    """Get milliseconds until expiration (None if perpetual, negative if expired)."""
    payload = extract_license_payload(license_string)
    if not payload or payload.exp is None:
        return None
    now_ms = int(time.time() * 1000)
    return payload.exp - now_ms
